File: ios/HaliSahaApp/HaliSahaApp/Utilities/Validators.swift
//
//  Validators.swift
//  HaliSahaApp
//
//  Form doğrulama yardımcı fonksiyonları
//
//  Created by Mehmet Mert Mazıcı on 23.12.2025.
//

import Foundation

// MARK: - Validator Protocol
protocol Validator {
    associatedtype Value
    func validate(_ value: Value) -> ValidationResult
}

// MARK: - Validation Result
struct ValidationResult {
    let isValid: Bool
    let errorMessage: String?
    
    static let valid = ValidationResult(isValid: true, errorMessage: nil)
    
    static func invalid(_ message: String) -> ValidationResult {
        ValidationResult(isValid: false, errorMessage: message)
    }
}

// MARK: - Email Validator
struct EmailValidator: Validator {
    func validate(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            return .invalid("E-posta adresi gerekli")
        }
        
        let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}"
        let predicate = NSPredicate(format: "SELF MATCHES %@", emailRegex)
        
        guard predicate.evaluate(with: trimmed) else {
            return .invalid("Geçerli bir e-posta adresi girin")
        }
        
        return .valid
    }
}

// MARK: - Password Validator
struct PasswordValidator: Validator {
    let minLength: Int
    let requireUppercase: Bool
    let requireNumber: Bool
    let requireSpecialChar: Bool
    
    init(
        minLength: Int = 6,
        requireUppercase: Bool = false,
        requireNumber: Bool = false,
        requireSpecialChar: Bool = false
    ) {
        self.minLength = minLength
        self.requireUppercase = requireUppercase
        self.requireNumber = requireNumber
        self.requireSpecialChar = requireSpecialChar
    }
    
    func validate(_ value: String) -> ValidationResult {
        guard !value.isEmpty else {
            return .invalid("Şifre gerekli")
        }
        
        guard value.count >= minLength else {
            return .invalid("Şifre en az \(minLength) karakter olmalı")
        }
        
        if requireUppercase {
            guard value.rangeOfCharacter(from: .uppercaseLetters) != nil else {
                return .invalid("Şifre en az bir büyük harf içermeli")
            }
        }
        
        if requireNumber {
            guard value.rangeOfCharacter(from: .decimalDigits) != nil else {
                return .invalid("Şifre en az bir rakam içermeli")
            }
        }
        
        if requireSpecialChar {
            let specialChars = CharacterSet(charactersIn: "!@#$%^&*()_+-=[]{}|;:,.<>?")
            guard value.rangeOfCharacter(from: specialChars) != nil else {
                return .invalid("Şifre en az bir özel karakter içermeli")
            }
        }
        
        return .valid
    }
}

// MARK: - Phone Validator (Türkiye)
struct PhoneValidator: Validator {
    func validate(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            return .invalid("Telefon numarası gerekli")
        }
        
        // Sadece rakamları al
        let digitsOnly = trimmed.components(separatedBy: CharacterSet.decimalDigits.inverted).joined()
        
        // Türkiye numarası: 5XX XXX XX XX (10 hane)
        // veya +90 5XX XXX XX XX (12 hane)
        // veya 0 5XX XXX XX XX (11 hane)
        
        let cleanNumber: String
        if digitsOnly.hasPrefix("90") && digitsOnly.count == 12 {
            cleanNumber = String(digitsOnly.dropFirst(2))
        } else if digitsOnly.hasPrefix("0") && digitsOnly.count == 11 {
            cleanNumber = String(digitsOnly.dropFirst(1))
        } else if digitsOnly.count == 10 {
            cleanNumber = digitsOnly
        } else {
            return .invalid("Geçerli bir telefon numarası girin")
        }
        
        // 5 ile başlamalı
        guard cleanNumber.hasPrefix("5") else {
            return .invalid("Telefon numarası 5 ile başlamalı")
        }
        
        return .valid
    }
}

// MARK: - Username Validator
struct UsernameValidator: Validator {
    let minLength: Int
    let maxLength: Int
    
    init(minLength: Int = 3, maxLength: Int = 30) {
        self.minLength = minLength
        self.maxLength = maxLength
    }
    
    func validate(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        
        guard !trimmed.isEmpty else {
            return .invalid("Kullanıcı adı gerekli")
        }
        
        guard trimmed.count >= minLength else {
            return .invalid("Kullanıcı adı en az \(minLength) karakter olmalı")
        }
        
        guard trimmed.count <= maxLength else {
            return .invalid("Kullanıcı adı en fazla \(maxLength) karakter olabilir")
        }
        
        // Sadece harf, rakam ve alt çizgi
        let allowedChars = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "_"))
        guard trimmed.unicodeScalars.allSatisfy({ allowedChars.contains($0) }) else {
            return .invalid("Kullanıcı adı sadece harf, rakam ve _ içerebilir")
        }
        
        // Rakam ile başlamamalı
        guard !trimmed.first!.isNumber else {
            return .invalid("Kullanıcı adı rakam ile başlayamaz")
        }
        
        return .valid
    }
}

// MARK: - Required Field Validator
struct RequiredValidator: Validator {
    let fieldName: String
    
    func validate(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            return .invalid("\(fieldName) gerekli")
        }
        
        return .valid
    }
}

// MARK: - Tax Number Validator (Türkiye)
struct TaxNumberValidator: Validator {
    func validate(_ value: String) -> ValidationResult {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        
        guard !trimmed.isEmpty else {
            return .invalid("Vergi numarası gerekli")
        }
        
        // Sadece rakamları al
        let digitsOnly = trimmed.components(separatedBy: CharacterSet.decimalDigits.inverted).joined()
        
        // 10 veya 11 hane olmalı
        guard digitsOnly.count == 10 || digitsOnly.count == 11 else {
            return .invalid("Vergi numarası 10 veya 11 hane olmalı")
        }
        
        return .valid
    }
}

// MARK: - Form Validator
struct FormValidator {
    
    static let email = EmailValidator()
    static let password = PasswordValidator()
    static let strongPassword = PasswordValidator(
        minLength: 8,
        requireUppercase: true,
        requireNumber: true
    )
    static let phone = PhoneValidator()
    static let username = UsernameValidator()
    static let taxNumber = TaxNumberValidator()
    
    static func required(_ fieldName: String) -> RequiredValidator {
        RequiredValidator(fieldName: fieldName)
    }
    
    // MARK: - Convenience Methods
    
    static func validateEmail(_ email: String) -> ValidationResult {
        self.email.validate(email)
    }
    
    static func validatePassword(_ password: String) -> ValidationResult {
        self.password.validate(password)
    }
    
    static func validatePhone(_ phone: String) -> ValidationResult {
        self.phone.validate(phone)
    }
    
    static func validateUsername(_ username: String) -> ValidationResult {
        self.username.validate(username)
    }
    
    static func validatePasswordMatch(_ password: String, _ confirm: String) -> ValidationResult {
        guard password == confirm else {
            return .invalid("Şifreler eşleşmiyor")
        }
        return .valid
    }
}

// MARK: - String Extension for Validation
extension String {
    
    var emailValidation: ValidationResult {
        FormValidator.validateEmail(self)
    }
    
    var passwordValidation: ValidationResult {
        FormValidator.validatePassword(self)
    }
    
    var phoneValidation: ValidationResult {
        FormValidator.validatePhone(self)
    }
    
    var usernameValidation: ValidationResult {
        FormValidator.validateUsername(self)
    }
}

File: android/app/src/main/java/com/example/HaliSahaApp/utils/Validators.kt
package com.example.HaliSahaApp.utils

// MARK: - Validation Result
data class ValidationResult(
    val isValid: Boolean,
    val errorMessage: String? = null
) {
    companion object {
        val Valid = ValidationResult(isValid = true, errorMessage = null)

        fun Invalid(message: String): ValidationResult {
            return ValidationResult(isValid = false, errorMessage = message)
        }
    }
}

// MARK: - Validator Interface
interface Validator<T> {
    fun validate(value: T): ValidationResult
}

// MARK: - Email Validator
class EmailValidator : Validator<String> {
    override fun validate(value: String): ValidationResult {
        val trimmed = value.trim()

        if (trimmed.isEmpty()) {
            return ValidationResult.Invalid("E-posta adresi gerekli")
        }

        // Android'in Patterns.EMAIL_ADDRESS'i yerine Swift ile aynı regex'i kullanıyoruz ki logic aynı olsun
        val emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}".toRegex(RegexOption.IGNORE_CASE)

        if (!emailRegex.matches(trimmed)) {
            return ValidationResult.Invalid("Geçerli bir e-posta adresi girin")
        }

        return ValidationResult.Valid
    }
}

// MARK: - Password Validator
class PasswordValidator(
    private val minLength: Int = 6,
    private val requireUppercase: Boolean = false,
    private val requireNumber: Boolean = false,
    private val requireSpecialChar: Boolean = false
) : Validator<String> {

    override fun validate(value: String): ValidationResult {
        if (value.isEmpty()) {
            return ValidationResult.Invalid("Şifre gerekli")
        }

        if (value.length < minLength) {
            return ValidationResult.Invalid("Şifre en az $minLength karakter olmalı")
        }

        if (requireUppercase && !value.any { it.isUpperCase() }) {
            return ValidationResult.Invalid("Şifre en az bir büyük harf içermeli")
        }

        if (requireNumber && !value.any { it.isDigit() }) {
            return ValidationResult.Invalid("Şifre en az bir rakam içermeli")
        }

        if (requireSpecialChar) {
            val specialChars = "!@#$%^&*()_+-=[]{}|;:,.<>?"
            if (!value.any { specialChars.contains(it) }) {
                return ValidationResult.Invalid("Şifre en az bir özel karakter içermeli")
            }
        }

        return ValidationResult.Valid
    }
}

// MARK: - Phone Validator (Türkiye)
class PhoneValidator : Validator<String> {
    override fun validate(value: String): ValidationResult {
        val trimmed = value.trim()

        if (trimmed.isEmpty()) {
            return ValidationResult.Invalid("Telefon numarası gerekli")
        }

        // Sadece rakamları al
        val digitsOnly = trimmed.filter { it.isDigit() }

        val cleanNumber = when {
            digitsOnly.startsWith("90") && digitsOnly.length == 12 -> digitsOnly.substring(2)
            digitsOnly.startsWith("0") && digitsOnly.length == 11 -> digitsOnly.substring(1)
            digitsOnly.length == 10 -> digitsOnly
            else -> return ValidationResult.Invalid("Geçerli bir telefon numarası girin")
        }

        if (!cleanNumber.startsWith("5")) {
            return ValidationResult.Invalid("Telefon numarası 5 ile başlamalı")
        }

        return ValidationResult.Valid
    }
}

// MARK: - Username Validator
class UsernameValidator(
    private val minLength: Int = 3,
    private val maxLength: Int = 30
) : Validator<String> {

    override fun validate(value: String): ValidationResult {
        val trimmed = value.trim().lowercase()

        if (trimmed.isEmpty()) {
            return ValidationResult.Invalid("Kullanıcı adı gerekli")
        }

        if (trimmed.length < minLength) {
            return ValidationResult.Invalid("Kullanıcı adı en az $minLength karakter olmalı")
        }

        if (trimmed.length > maxLength) {
            return ValidationResult.Invalid("Kullanıcı adı en fazla $maxLength karakter olabilir")
        }

        // Sadece harf, rakam ve alt çizgi
        if (!trimmed.all { it.isLetterOrDigit() || it == '_' }) {
            return ValidationResult.Invalid("Kullanıcı adı sadece harf, rakam ve _ içerebilir")
        }

        // Rakam ile başlamamalı
        if (trimmed.first().isDigit()) {
            return ValidationResult.Invalid("Kullanıcı adı rakam ile başlayamaz")
        }

        return ValidationResult.Valid
    }
}

// MARK: - Required Field Validator
class RequiredValidator(private val fieldName: String) : Validator<String> {
    override fun validate(value: String): ValidationResult {
        if (value.trim().isEmpty()) {
            return ValidationResult.Invalid("$fieldName gerekli")
        }
        return ValidationResult.Valid
    }
}

// MARK: - Tax Number Validator
class TaxNumberValidator : Validator<String> {
    override fun validate(value: String): ValidationResult {
        val trimmed = value.trim()

        if (trimmed.isEmpty()) {
            return ValidationResult.Invalid("Vergi numarası gerekli")
        }

        val digitsOnly = trimmed.filter { it.isDigit() }

        if (digitsOnly.length != 10 && digitsOnly.length != 11) {
            return ValidationResult.Invalid("Vergi numarası 10 veya 11 hane olmalı")
        }

        return ValidationResult.Valid
    }
}

// MARK: - Form Validator Singleton
object FormValidator {

    val email = EmailValidator()
    val password = PasswordValidator()
    val strongPassword = PasswordValidator(
        minLength = 8,
        requireUppercase = true,
        requireNumber = true
    )
    val phone = PhoneValidator()
    val username = UsernameValidator()
    val taxNumber = TaxNumberValidator()

    fun required(fieldName: String) = RequiredValidator(fieldName)

    // MARK: - Convenience Methods

    fun validateEmail(email: String): ValidationResult = this.email.validate(email)

    fun validatePassword(password: String): ValidationResult = this.password.validate(password)

    fun validatePhone(phone: String): ValidationResult = this.phone.validate(phone)

    fun validateUsername(username: String): ValidationResult = this.username.validate(username)

    fun validatePasswordMatch(password: String, confirm: String): ValidationResult {
        return if (password == confirm) {
            ValidationResult.Valid
        } else {
            ValidationResult.Invalid("Şifreler eşleşmiyor")
        }
    }
}

// MARK: - String Extensions for Validation
val String.emailValidation: ValidationResult
    get() = FormValidator.validateEmail(this)

val String.passwordValidation: ValidationResult
    get() = FormValidator.validatePassword(this)

val String.phoneValidation: ValidationResult
    get() = FormValidator.validatePhone(this)

val String.usernameValidation: ValidationResult
    get() = FormValidator.validateUsername(this)